src/Storages/PostgreSQL/StoragePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,514 @@
+        interpreter.execute();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    nested_storage = getNested();
+}
+
+
+Context StoragePostgreSQLReplica::makeNestedTableContext() const
+{
+    auto context(*global_context);
+    context.makeQueryContext();
+    context.addQueryFactoriesInfo(Context::QueryLogFactories::Storage, "ReplacingMergeTree");
+
+    return context;
+}
+
+
+StoragePtr StoragePostgreSQLReplica::getNested()
+{
+    if (nested_storage)
+        return nested_storage;
+
+    auto context = makeNestedTableContext();
+    nested_storage = DatabaseCatalog::instance().getTable(
+            StorageID(getStorageID().database_name, getNestedTableName()), context);
+
+    return nested_storage;
+}
+
+
+StoragePtr StoragePostgreSQLReplica::tryGetNested()
+{
+    if (nested_storage)
+        return nested_storage;
+
+    auto context = makeNestedTableContext();
+    nested_storage = DatabaseCatalog::instance().tryGetTable(
+            StorageID(getStorageID().database_name, getNestedTableName()), context);
+
+    return nested_storage;
+}
+
+
+void StoragePostgreSQLReplica::startup()
+{
+    if (!is_postgresql_replica_database)
+    {
+        replication_handler->addStorage(remote_table_name, this);
+        replication_handler->startup();
+    }
+}
+
+
+void StoragePostgreSQLReplica::shutdown()
+{
+    if (replication_handler)
+        replication_handler->shutdown();
+}
+
+
+void StoragePostgreSQLReplica::shutdownFinal()
+{
+    if (is_postgresql_replica_database)
+        return;
+
+    if (replication_handler)
+        replication_handler->shutdownFinal();
+
+    if (nested_storage)
+        dropNested();
+}
+
+
+void StoragePostgreSQLReplica::dropNested()
+{
+    std::lock_guard lock(nested_mutex);
+    nested_loaded = false;
+
+    auto table_id = nested_storage->getStorageID();
+    auto ast_drop = std::make_shared<ASTDropQuery>();
+
+    ast_drop->kind = ASTDropQuery::Drop;
+    ast_drop->table = table_id.table_name;
+    ast_drop->database = table_id.database_name;
+    ast_drop->if_exists = true;
+
+    auto context = makeNestedTableContext();
+    auto interpreter = InterpreterDropQuery(ast_drop, context);
+    interpreter.execute();
+
+    nested_storage = nullptr;
+    LOG_WARNING(&Poco::Logger::get("StoragePostgreSQLReplica"), "Dropped (or temporarily) nested table {}", getNestedTableName());
+}
+
+
+NamesAndTypesList StoragePostgreSQLReplica::getVirtuals() const
+{
+    if (nested_storage)
+        return nested_storage->getVirtuals();
+
+    return {};
+}
+
+
+Pipe StoragePostgreSQLReplica::read(
+        const Names & column_names,
+        const StorageMetadataPtr & /* metadata_snapshot */,
+        SelectQueryInfo & query_info,
+        const Context & context,
+        QueryProcessingStage::Enum processed_stage,
+        size_t max_block_size,
+        unsigned num_streams)
+{
+    std::unique_lock lock(nested_mutex, std::defer_lock);
```

Comment:
Why defer_lock if we try to lock it on the next line?

Reply:
Removed it completely.